core/domain/stats_services_test.py

```diff
@@ -246,6 +251,33 @@ def test_handle_stats_creation_for_new_exp_version(self):
             exploration_stats.state_stats_mapping.keys(), [
                 'Home', 'New state 4', 'Renamed state', 'End'])
 
+        # Test deletion, addition and rename.
+        exploration.delete_state('New state 4')
+        exploration.add_states(['New state'])
+        exploration.rename_state('New state', 'New state 4')
+        exploration.version += 1
+        change_list = [{
+            'cmd': 'delete_state',
+            'state_name': 'New state 4'
+        }, {
+            'cmd': 'add_state',
+            'state_name': 'New state',
+        }, {
+            'cmd': 'rename_state',
+            'old_state_name': 'New state',
+            'new_state_name': 'New state 4'
+        }]
+        stats_services.handle_stats_creation_for_new_exp_version(
+            exploration.id, exploration.version, exploration.states,
+            change_list)
+
+        exploration_stats = stats_services.get_exploration_stats_by_id(
+            exploration.id, exploration.version)
+        self.assertEqual(exploration_stats.exp_version, 7)
+        self.assertEqual(
+            exploration_stats.state_stats_mapping.keys(), [
+                'Home', 'New state 4', 'Renamed state', 'End'])
```

Comment:
You need to check more than this -- make sure the stats have been correctly carried over in a way that preserves "state identity". Ditto above.

Reply:
@seanlip Done. I've added a test (Lines 254 onward) which sets values for exploration stats model and later tests how they get carried over after a version change.